### PR TITLE
Active record pattern

### PR DESCRIPTION
**Pull request by TwentyVentti**
Pull request title

- Attempting to implement the active record pattern in java.
- https://github.com/iluwatar/java-design-patterns/issues/79

> For detailed contributing instructions see https://github.com/iluwatar/java-design-patterns/wiki/01.-How-to-contribute

### CODE REVIEW

File: active-record/README.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
+---
+title: Active Record
```
Comment (Jeremy-Cains-ANU) [inline active-record/README.md:2]: If you have a look at newer patterns, there is more info in the table. Here is Hugo's
```---
layout: pattern
title: Currying
folder: currying
permalink: /patterns/currying/
categories: Functional
language: en
tags: 
 - Decoupling
---
```

File: active-record/etc/active-record.urm.puml [hunk 2]
```diff
@@ -0,0 +1,40 @@
+@startuml
```
Comment (Jeremy-Cains-ANU) [inline active-record/etc/active-record.urm.puml:1]: You need to upload the image generated by this yourself. There is info about it on the wiki.

File: active-record/src/main/java/com/iluwatar/activeobject/ActiveRow.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+  @Rowverride
+  public void write() {
+    StringBuilder query = new StringBuilder();
+    query.append(write);
+    query.append(" VALUES (");
+    for (String con : this.contents) {
+      if (contents.indexOf(con) != this.contents.size() - 1) {
+        query.append("'");
+        query.append(con);
+        query.append("' ");
```
Comment (Jeremy-Cains-ANU) [inline active-record/src/main/java/com/iluwatar/activeobject/ActiveRow.java]: Some of these lines could be condenses, it seems

File: active-record/src/test/java/com/iluwatar/activeobject/ActiveRowTest.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+	}
+
+	@Test
+	void insertRowTest() {
+		ActiveDatabase activeDatabase = new ActiveDatabase("world", "root", "apple-trunks", "localhost:3306", "city");
+		ActiveRow row = new ActiveRow(activeDatabase, "1");
+		row.delete();
+		row.contents = new ArrayList<String>(
+				Arrays.asList("101", "Godoy Cruz", "ARG", "Mendoza", "206998"));
+		assertEquals(row.read(), Arrays.asList("101", "Godoy Cruz", "ARG", "Mendoza", "206998"));
```
Comment (Jeremy-Cains-ANU) [inline active-record/src/test/java/com/iluwatar/activeobject/ActiveRowTest.java]: You delete the row and then assign it's contents, so I don't think it's fair to call it the insertRowTest. What would deleted rows 29-50? Isn't that the test you want?

File: active-record/src/main/java/com/iluwatar/activeobject/Rowverride.java [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Target(ElementType.METHOD)
+@Retention(RetentionPolicy.SOURCE)
+public @interface Rowverride {
+
+}
```
Comment (Jeremy-Cains-ANU) [inline active-record/src/main/java/com/iluwatar/activeobject/Rowverride.java:36]: What is this for?

File: active-record/README.md [hunk 6]
```diff
@@ -0,0 +1,15 @@
+---
+title: Active Record
+category: Architectural  
+language: en
+tags:
+ - Data access
+---
+
+## Intent
+The active record design pattern is a way of accessing data from databases by creating an object which contains the contents of a database row.
```
Comment (Jeremy-Cains-ANU) [inline active-record/README.md:10]: There's a bit more to it, isn't there? Like each row having Insert, Update, and Delete

**Review by Jeremy-Cains-ANU — CHANGES_REQUESTED** [review]

File: active-record/src/main/java/com/iluwatar/activeobject/ActiveRow.java (hunk 3 above)
Comment (TwentyVentti) [reply active-record/src/main/java/com/iluwatar/activeobject/ActiveRow.java]: True, like query.append("'").append(con).append("' ");?

File: active-record/src/test/java/com/iluwatar/activeobject/ActiveRowTest.java (hunk 4 above)
Comment (TwentyVentti) [reply active-record/src/test/java/com/iluwatar/activeobject/ActiveRowTest.java]: Good catch, yep I meant to row.delete() then update contents, then write it and then read().

File: active-record/src/main/java/com/iluwatar/activeobject/Rowverride.java (hunk 5 above)
Comment (TwentyVentti) [reply active-record/src/main/java/com/iluwatar/activeobject/Rowverride.java:36]: Good question, its an interface used to override static methods.

File: active-record/README.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+language: en
+tags:
+ - Data access
+---
+
+## Intent
+The active record design pattern is a way of accessing data from databases by creating an object which contains the contents of a database row.
+
+## Explanation
+The ActiveRecord object is intitialised by an ActiveDatabase object which creates a connection to an exisitng database.
```
Comment (Jesonzavic) [inline active-record/README.md:13]: This explanation is not sufficient. Please see other examples in the repo for what is to be expected.

File: active-record/README.md [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+
+**Programmatic Example**
+
+This implementation shows what the Active Record pattern could look like in a generic context. The Active Record makes initialises `contents` and `columns` as lists of strings. A developer can create an `ActiveDatabase` object by entering the configuration details required to start the connection. Assuming there is an exsiting database the developer would then create an `ActiveRecord` object using the `ActiveDatabase` object and the `id` of the row with which they would like to create in an active context. They are able to locate the row which matches this `id` then this can be read or deleted. 
+
+If you are adding in a new row you can just update the `contents` list which would be empty if created with an `id` which doesnt exist yet.
```
Comment (Jesonzavic) [inline active-record/README.md:31]: You are missing the java example portions of the programmatic example. Please see any other pattern for how it should be structured.

File: active-record/etc/active-record.urm.puml [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+  }
+  class App {
+    + App()
+    + main(args : String[]) {static}
+  }
+  interface Rowverride {
+  }
+}
+ActiveRow -->  "-dataBase" ActiveDatabase
+@enduml
```
Comment (Jesonzavic) [inline active-record/etc/active-record.urm.puml:40]: You are missing: Applicability, Consequences, Related Patterns, and Credits in this readme.

**Review by Jesonzavic — CHANGES_REQUESTED** [review]
Just a review of the README for now, adding onto what has already been commented.

**Pull request merged**